src/examples/pmreorder/pmreorder_list.c

```diff
@@ -0,0 +1,186 @@
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * pmreorder_list.c -- explains how to use pmreorder tool with libpmem
+ *
+ * usage: pmreorder_list <g|b|c> <path>
+ * g - good case - add element to the list in a consistent way
+ * b - bad case - add element to the list in an inconsistent way
+ * c - check persistent consistency of the list
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <libpmem.h>
+
+#define MAX_NODES 10
+#define NODE_PTR(root, node_id) (&(root)->nodes[(node_id)])
+
+typedef size_t node_id;
+
+struct list_node {
+	int value;
+	node_id next;
+};
+
+struct list_root {
+	node_id head;
+	struct list_node nodes[MAX_NODES];
+};
+
+/*
+ * check_consistency -- check if list meets the set requirements
+ *
+ * for consistent cases function returns 0
+ * for inconsistent cases function returns 1
+ */
+static int
+check_consistency(struct list_root *root)
+{
+	/* if all nodes are set... */
+	if (NODE_PTR(root, 5)->value != 0 &&
+			NODE_PTR(root, 3)->value != 0 &&
+			NODE_PTR(root, 6)->value != 0) {
+		/* ...then we expect they have a proper order and values */
+		struct list_node *node1 = NODE_PTR(root, root->head);
+		struct list_node *node2 = NODE_PTR(root, node1->next);
+		struct list_node *node3 = NODE_PTR(root, node2->next);
+
+		if (node1->value != 66 ||
+			node2->value != 33 ||
+			node3->value != 55)
+			return 1;
+	}
+	return 0;
+}
+
+/*
+ * list_print -- prints all  elements to the screen
+ */
+static void
+list_print(struct list_root *list)
+{
+	printf("List:\n");
+	node_id n = list->head;
+	while (NODE_PTR(list, n)->value != 0) {
+		printf("Value: %d\n", NODE_PTR(list, n)->value);
+		n = NODE_PTR(list, n)->next;
+	}
+}
+
+/*
+ * list_insert_consistency -- add new element to the list in a consistent way
+ */
+static void
+list_insert_consistent(struct list_root *root, node_id node, int value)
+{
+	struct list_node *new = NODE_PTR(root, node);
+
+	new->next = root->head;
+	pmem_persist(&new->next, sizeof(node));
+
+	root->head = node;
+	pmem_persist(&root->head, sizeof(root->head));
+
+	new->value = value;
+	pmem_persist(&new->value, sizeof(value));
```

Comment:
This is not correct. root->head update needs to be the last operation. Otherwise node is visible before it's fully initialized.